rero_ils/modules/items/utils.py

```diff
@@ -85,3 +88,28 @@ def exists_available_item(items=[]):
         if item.available:
             return True
     return False
+
+
+def get_provisional_items_pids_candidate_to_delete():
+    """Returns checked-in provisional items pids.
+
+    Returns list of candidate provisional items pids to delete, based on the
+    status of the item. Filtering by the status `ItemStatus.ON_SHELF` removes
+    items with active loans. in addition, remove items with active fees.
+    :return an item pid generator.
+    """
+    from rero_ils.modules.items.api import ItemsSearch
+    query_fees = PatronTransactionsSearch()\
+        .filter('term', status='open')\
+        .filter('range', total_amount={'gt': 0})\
+        .source('item')
+    item_pids_with_fees = list(
+        set([hit.item.pid for hit in query_fees.scan()]))
```

Comment:
instead of building yourself the set of item_pids, it should be more powerfull to ask ES to build it using an aggregation (facet) on `item.pid`.

You could inspired by code from `holdings/listener.pyL41` ::
```python
query_fees.aggs.bucket('item_pids', 'terms', field='item.pid')
items_pids_with_fees = results.aggregations.item_pids.buckets. # check the structure of returned bucket.
```

Reply:
c'est fait